addon/-private/class/modifier.ts

```diff
@@ -64,6 +85,55 @@ export default class ClassBasedModifier<S = DefaultSignature> {
     deprecateForDestroyables('willDestroy', this);
     deprecateForDestroyables('isDestroying', this);
     deprecateForDestroyables('isDestroyed', this);
+
+    assert(
+      'ember-modifier: You cannot implement both `run` and any of the deprecated legacy lifecycle hooks (`didInstall`, `didReceiveArguments`, and `didUpdateArguments`)',
+      !(_implementsModify(this) && _implementsLegacyHooks(this))
+    );
+  }
+
+  /**
+   * Called when the modifier is installed and any time any tracked state used
+   * in the modifier changes.
+   *
+   * If you need to do first-time-only setup, create a class field representing
+   * the initialization state and check it when running the hook. That is also
+   * where and when you should use `registerDestructor` for any teardown you
+   * need to do. For example:
+   *
+   * ```js
+   * function disconnect(instance) {
+   *  instnace.observer?.disconnect();
+   * }
+   *
+   * class IntersectionObserver extends Modifier {
+   *   observer;
+   *
+   *   modify(element, callback, options) {
+   *     // Gets rid of any existing connection, since we're getting called with
+   *     // a new callback or new options.
+   *     disconnect(this);
+   *
+   *     this.observer = new IntersectionObserver(callback, options);
+   *     this.observer.observe(element);
+   *
+   *     registerDestructor(this, disconnect);
```

Comment:
FWIW, I think this would register a new destructor on every mutation (never culling/removing the prior value)

Reply:
Ah, good catch. Yeah, I’ll update the example.

Reply:
Updated in ef0a76c – let me know if that *isn’t* the right way to do it!